server/lib/src/idm/ldap.rs

```diff
@@ -525,6 +532,107 @@ impl LdapServer {
             },
         } // end match server op
     }
+
+    fn naming_context_from_base_dn(&self, _dn: &str) -> Result<&LdapNamingContext, OperationError> {
+        Ok(&self.default_naming_context)
+    }
+
+    async fn bind_target_from_bind_dn(
+        &self,
+        idms: &IdmServer,
+        dn: &str,
+        pw: &str,
+    ) -> Result<LdapBindTarget, OperationError> {
+        if dn.is_empty() {
+            if pw.is_empty() {
+                return Ok(LdapBindTarget::Account(UUID_ANONYMOUS));
+            } else {
+                // This is the path to access api-token logins.
+                return Ok(LdapBindTarget::ApiToken);
+            }
+        } else if dn == "dn=token" {
+            // Is the passed dn requesting token auth?
+            // We use dn= here since these are attr=value, and dn is a phantom so it will
+            // never be present or match a real value. We also make it an ava so that clients
+            // that over-zealously validate dn syntax are happy.
+            return Ok(LdapBindTarget::ApiToken);
+        }
+
+        if let Some(rdn) = self
+            .default_naming_context
+            .binddnre
+            .captures(dn)
+            .and_then(|caps| caps.name("val"))
+            .map(|v| v.as_str().to_string())
+        {
+            if rdn.is_empty() {
+                // That's weird ...
+                return Err(OperationError::NoMatchingEntries);
+            }
+
+            let mut idm_auth = idms.auth().await;
+            let uuid = idm_auth.qs_read.name_to_uuid(rdn.as_str()).map_err(|e| {
+                request_error!(err = ?e, ?rdn, "Error resolving rdn to target");
+                e
+            })?;
+
+            return Ok(LdapBindTarget::Account(uuid));
+        }
+
+        let basedn = &self.default_naming_context.basedn;
+        let app_binddnre = Regex::new(
+            format!("^(([^=,]+)=)(?P<usr>[^=,]+),(([^=,]+)=)(?P<app>[^=,]+)(,{basedn})$").as_str(),
+        )
+        .map_err(|_| OperationError::InvalidEntryState)?;
+
+        if let Some(caps) = app_binddnre.captures(dn) {
+            let app_rdn = caps
+                .name("app")
+                .map(|v| v.as_str().to_string())
+                .ok_or(OperationError::NoMatchingEntries)?;
+            let usr_rdn = caps
+                .name("usr")
+                .map(|v| v.as_str().to_string())
+                .ok_or(OperationError::NoMatchingEntries)?;
+            if app_rdn.is_empty() || usr_rdn.is_empty() {
+                return Err(OperationError::NoMatchingEntries);
+            }
+
+            let mut idms_prox_read = idms.proxy_read().await;
+            let app_entries = idms_prox_read.qs_read.internal_search(filter!(f_eq(
+                Attribute::Class,
+                EntryClass::Application.into()
+            )))?;
+
+            for app_entry in &app_entries {
```

Comment:
We shouldn't need this loop, this is another place where we would use the ldap application cache so we can look up applications by their name from a hashmap.